pkg/mutation/base_mutator.go

```diff
@@ -0,0 +1,122 @@
+package mutation
+
+import (
+	"fmt"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/path/parser"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func NewMutator(path parser.Path, value string) BaseMutator {
+	return BaseMutator{
+		path:  path,
+		Value: value,
+	}
+}
+
+// Base is a base mutator wrapping the Assign and AssignMetadata type.
+type BaseMutator struct {
+	path  parser.Path
+	Value string
+}
+
+func (m BaseMutator) Mutate(obj *unstructured.Unstructured) error {
+	return m.mutate(obj.Object, nil, 0)
+}
+
+func (m BaseMutator) mutate(current interface{}, previous interface{}, depth int) error {
+	pathEntry := m.Path().Nodes[depth]
+	if len(m.Path().Nodes)-1 == depth {
+		current.(map[string]interface{})[pathEntry.(parser.Object).Reference] = m.Value
+		return nil
+	}
+	switch t := pathEntry.Type(); t {
+	case parser.ObjectNode:
+		next, ok := current.(map[string]interface{})[pathEntry.(parser.Object).Reference]
+		if !ok {
+			next = m.createMissingElement(current, previous, depth)
+		}
+		if err := m.mutate(next, current, depth+1); err != nil {
+			return err
+		}
+		return nil
+	case parser.ListNode:
+		elementFound := false
+		for _, listElement := range current.([]interface{}) {
+			if pathEntry.(parser.List).Glob {
+				if err := m.mutate(listElement, current, depth+1); err != nil {
+					return err
+				}
+				elementFound = true
+			} else if elementValue, ok := listElement.(map[string]interface{})[pathEntry.(parser.List).KeyField]; ok {
+				if *pathEntry.(parser.List).KeyValue == elementValue {
+					if err := m.mutate(listElement, current, depth+1); err != nil {
+						return err
+					}
+					elementFound = true
+				}
+			}
+		}
+
+		// If no matching element in the array was found in non Globbed list, create a new element
+		if !pathEntry.(parser.List).Glob && !elementFound {
+			next := m.createMissingElement(current, previous, depth)
+			if err := m.mutate(next, current, depth+1); err != nil {
+				return err
+			}
+		}
+	default:
+		return fmt.Errorf("Unrecognized type: %v", t)
+	}
+	return nil
+}
+
+func (m BaseMutator) createMissingElement(current interface{}, previous interface{}, depth int) interface{} {
+	var next interface{}
+	pathEntry := m.Path().Nodes[depth]
+
+	// Create new element of type
+	switch m.Path().Nodes[depth+1].Type() {
+	case parser.ObjectNode:
+		next = make(map[string]interface{})
+	case parser.ListNode:
+		next = make([]interface{}, 0)
+	}
+
+	// Append to element of type
+	switch m.Path().Nodes[depth].Type() {
+	case parser.ObjectNode:
+		current.(map[string]interface{})[pathEntry.(parser.Object).Reference] = next
+	case parser.ListNode:
+		current = append(current.([]interface{}), next)
+		next.(map[string]interface{})[pathEntry.(parser.List).KeyField] = *pathEntry.(parser.List).KeyValue
+		previous.(map[string]interface{})[m.Path().Nodes[depth-1].(parser.Object).Reference] = current
```

Comment:
I'm curious if this line is necessary? It may just work without explicit assignment

Reply:
append does not modify in place but returns a new object, therefore `current` in line 93 becomes a new object, and is not attached anywhere.
`previous.(map[string]interface{})[m.Path().Nodes[depth-1] ` will still point to the old `current`. We have to manully attach it.

I added a test to validate this: [here](https://github.com/open-policy-agent/gatekeeper/pull/937/files#diff-6e2f011a49c787ceb82484c404f0a75983ee85614f8f21e1191cda9c5ee9ee22R192)
If you remove the line the test will fail.

Reply:
Good point, thank you for looking into this.